### ISSUE TITLE
Timing/Profiling

### ISSUE THREAD

**Issue by russelljjarvis** [russelljjarvis/neuronunit#1]
Hi @rgerkin
Time 
`Jneuroml->pyneuroml->NRN=4.6258630752593355` seconds
`NRN=1.054311752319336` seconds
the time difference is: 
`3.57155132294` seconds

This is just using the time module not the line_profiler that you recommended to use which i have installed with the docker command here:
```
RUN pip install git+https://github.com/rkern/line_profiler
```
I have created a branch of my personal version of neuronunit called 'before_holidays'

If you run the file [Optomize.py](https://github.com/russelljjarvis/neuronunit/blob/before_holidays/neuronunit/software_tests/Optimize.py):

This file also performs a 5 parameter GA optimization using the NSGA algorithm and NEURON in serial.

You just need to press 'c' after running it to get past the point where python prints out the time difference between JNeuroML and NEURON

You will see where these time differences come from. Note this is not yet mean time differences (taking into account more complex firing patterns of the membrane potential. I think its safe to assume that the performance hit coming from compiling NMODL write a JNeuroML file and translating JNeuroml to NEURON is relatively constant however.

Over the holidays 
I was trying to finish the backends.py file that you described, such that it has a JNeuroML backend and NEURON backend. This proved more difficult than I expected.

Here is the current state of the [backends](https://github.com/russelljjarvis/neuronunit/blob/2e7ea474924096b748d52462e3c47fea0384c4b4/neuronunit/models/backends.py) file. Also note I probably need help with this file.

I am at the stage now where I can work on plotting the outputs of the optimization to check if they are sensible. I probably should try to make it so that the code runs on GH via web hooks instead of on my local machine.

**Comment by russelljjarvis** [comment]
Hi @rgerkin
![evolved_pop](https://cloud.githubusercontent.com/assets/7786645/21994905/4c4c0f14-dbdf-11e6-8272-e247b790dc1a.png)

Also the total optimization routine is taking approximately 4 minutes and it uses 24 invocations of the neuron simulator `NGEN=3`, `population size=8`.

**Comment by russelljjarvis** [comment]
For 
```
NGEN=4
MU=12
```
The total run time is: `14.966` `~15` minutes. This is not simply `4*12=48` invocations of NEURON since each search for Rheobase consists of many serial calls to NEURON, so the actual 

The mean call length for NEURON in a simulation that involves spiking is `2.08561599255` seconds

**Comment by rgerkin** [comment]
Are you changing the range of the parameters to match what you think are likely to be the locations of the optima (i.e. you know that searching outside that range is a waste of time)?  If so, future optimizations for different models or neuron types are likely to take longer because we don't yet know which locations are uninteresting.

**Comment by russelljjarvis** [comment]
So my ranges are currently the same as the ipython notebook AIBS.ipynb
```
vr = np.linspace(-75.0,-50.0,1000)
a = np.linspace(0.015,0.045,1000)
b = np.linspace(-3.5*10E-9,-0.5*10E-9,1000)
```
At the moment the optimizer raises an error if no spiking occurs, or basically when the score contains an insufficient data result.

I am attempting to make the code more robust, by catching this exception and filling up the error components with a nominally high value `-100,-100,...,-100,` etc. However this does not work in the parallel case and I am not sure why, but I would like to do more work on this issue, as it limits searches to parameter spaces that only return spiking values (score is of float type).

Reading through this thread do you feel like you have what you need in terms of the speed up?

```
    try:
        individual.error = []
        individual.error = [ abs(i.score) for i in score.unstack() ]
        individual.s_html=score.to_html()
    except Exception as e:
        '{}'.format('Insufficient Data')
        individual.error = []
        individual.error = [ 100.0 for i in range(0,7) ]
        if len(LOCAL_RESULTS_spiking)>0:
            del LOCAL_RESULTS_spiking[-1]
        individual.s_html=None
```

**Comment by rgerkin** [comment]
This is possibly a terrible idea but if the cliff (going from some normal fitness value to -100) is causing problems do you think it is possible to smooth it by averaging the -100 with the value stored for some adjacent parameter values?  Totally not even worth looking into it unless that -100 cliff is causing problems.

**Comment by rgerkin** [comment]
Counting the number of simulator calls from within `sciunit.Suite` (or `sciunit.Test`) would be good.  But don't count the number of simulator calls directly, just the number of calls to `judge`.  Each call to judge may correspond to multiple simulator calls, but that's fine.

**Comment by rgerkin** [comment]
As for documenting the speedup, could you fill in your estimates of the following values, based on what you wrote above or otherwise know:  

Speedup from parallelization is ___.
Speedup from modifying NEURON in memory vs writing new NeuroML files is ___.
Other sources of speedup are ____ and they produce a speedup of ____.

**Comment by russelljjarvis** [comment]
Okay so I will fill out the speed ups, and then I will get back to the issue of fixing condition when no spiking occurs.

**Comment by russelljjarvis** [comment]
I have created a [parallel](https://github.com/russelljjarvis/neuronunit/blob/compare_justas/neuronunit/tests/brute_force.py) method for exhaustive searching so at least I will have a table to compare with.
It can run in serial too. It relies in MPI which I have factored out.

I will can use my replies above to fill out:

Speedup from parallelization is ___.
Speedup from modifying NEURON in memory vs writing new NeuroML files is ___.
Other sources of speedup are ____ and they produce a speedup of ____.

Except other sources of speed up might include parallelizing the rheobase search. Using the `numba` just in time compiler for numpy arrays, writing code directly in cython.

I have not done any of these things. I think its better to stick with what we have and move towards BluePyOpt integration and the NSG cluster. These things will take time too. 

We have to way up debugging time/ versus speed up time. Any of the extra speed ups will doubtlessly add to the complexity, where as I think given we have achieved some level of parallelization and speed up now we should work on code robustness, maintainability and simplicity. 

Each time I make some minor changes to the code there is a huge debugging overhead. So some of these timing and profiling tests, are kind of 'already there'. The minor changes that I introduce are kind of costly in terms of the 15 minute wait time, so I am less inclined to work on future speed ups, and more inclined to document the existing speed ups and work on solution quality (and thus the exhaustive search method above to provide a ground truth, error surface).
